gcp-pubsub/src/main/java/io/micronaut/gcp/pubsub/intercept/PubSubConsumerAdvice.java

```diff
@@ -0,0 +1,171 @@
+                .getAnnotationTypeByStereotype(Qualifier.class)
+                .map(type -> Qualifiers.byAnnotation(beanDefinition, type))
+                .orElse(null);
+        boolean hasAckArg = Arrays.stream(method.getArguments())
+                .anyMatch(arg -> Acknowledgement.class.isAssignableFrom(arg.getType()));
+
+        Class<Object> beanType = (Class<Object>) beanDefinition.getBeanType();
+        Object bean = beanContext.findBean(beanType, qualifer).orElseThrow(() -> new MessageListenerException("Could not find the bean to execute the method " + method));
+        DefaultExecutableBinder<PubSubConsumerState> binder = new DefaultExecutableBinder<>();
+
+        if (subscriptionAnnotation != null) {
+            String subscriptionName = subscriptionAnnotation.getRequiredValue(String.class);
+            ProjectSubscriptionName projectSubscriptionName = PubSubSubscriptionUtils.toProjectSubscriptionName(subscriptionName, googleCloudConfiguration.getProjectId());
+            String defaultContentType = subscriptionAnnotation.get("contentType", String.class).orElse("");
+            MessageReceiver receiver = (PubsubMessage message, AckReplyConsumer consumer) -> {
+                String messageContentType = message.getAttributesMap().getOrDefault("Content-Type", "");
+                String contentType = Optional.of(messageContentType).orElse(defaultContentType);
+                DefaultPubSubAcknowledgement pubSubAcknowledgement = new DefaultPubSubAcknowledgement(consumer);
+
+                PubSubConsumerState consumerState = new PubSubConsumerState(message, consumer,
+                        projectSubscriptionName, contentType);
+                try {
+                    BoundExecutable executable = null;
+                    try {
+                        executable = binder.bind(method, binderRegistry, consumerState);
+                    } catch (Exception ex) {
+                        handleException(new PubSubMessageReceiverException("Error binding message to the method", ex, bean, consumerState));
+                    }
+                    executable.invoke(bean); // Discard result
+                    if (!hasAckArg) { // if manual ack is not specified we auto ack message after method execution
+                        pubSubAcknowledgement.ack();
+                    } else {
+                        if (!pubSubAcknowledgement.isMessageAcked()) {
+                            logger.warn("Method {} was executed and no message acknowledge detected, message will be redelivered.", method.getName());
+                        }
+                    }
+                } catch (Exception e) {
+                    handleException(new PubSubMessageReceiverException("Error handling message", e, bean, consumerState));
+                }
+            };
+            try {
+                lock.lock();
+                if (registeredSubscribers.containsKey(projectSubscriptionName.toString())) {
```

Comment:
In general here I think you have the concern of multiple subscribers for a single subscription duplicated here and in the subscriber factory. It should just be in one place. I think the factory should be responsible for creating the subscriber, throwing the exception if it already exists, caching the result, and cleaning it up at shutdown

Reply:
yes, somehow I forgot that the Factory was already handling this, cleaned this up.